File: liqp-core/src/test/java/liqp/filter/MinusTest.java
package liqp.filter;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import liqp.LiquidDefaults;
import liqp.LiquidParser;
import liqp.Mocks;
import liqp.LiquidTemplate;
import org.antlr.runtime.RecognitionException;
import org.assertj.core.api.Assertions;
import org.junit.Test;

public class MinusTest {

  @Test
  public void applyTest() throws RecognitionException {

    String[][] tests = {
          {"{{ 8 | minus: 2 }}", "6"},
          {"{{ 8 | minus: 3 }}", "5"},
          {"{{ 8 | minus: 3. }}", "5.0"},
          {"{{ 8 | minus: 3.0 }}", "5.0"},
          {"{{ 8 | minus: 2.0 }}", "6.0"},
    };

    for (String[] test : tests) {

      LiquidTemplate template = LiquidParser.newInstance().parse(test[0]);
      String rendered = String.valueOf(template.render());

      assertThat(rendered, is(test[1]));
    }
  }

  @Test
  public void applyTestInvalid1() {
    Assertions.assertThat(
          LiquidDefaults.getDefaultFilters().getFilter("minus").onFilterAction(Mocks.mockRenderContext(), 1))
          .isEqualTo(1);
  }

  @Test
  public void applyTestInvalid2() {
    Assertions.assertThat(
          LiquidDefaults.getDefaultFilters().getFilter("minus").onFilterAction(Mocks.mockRenderContext(), 1, 2, 3))
          .isEqualTo(-4L);
  }

  /*
   * def test_minus
   *   assert_template_result "4", "{{ input | minus:operand }}", 'input' => 5, 'operand' => 1
   *   assert_template_result "2.3", "{{ '4.3' | minus:'2' }}"
   * end
   */
  @Test
  public void applyOriginalTest() {

    assertThat(LiquidParser.newInstance().parse("{{ input | minus:operand }}").render("{\"input\":5, \"operand\":1}")
          , is((Object) "4"));
    assertThat(LiquidParser.newInstance().parse("{{ '4.3' | minus:'2' }}").render(), is((Object) "2.3"));
  }
}

File: liqp-core/src/test/kotlin/liqp/filter/MinusTest.kt
package liqp.filter

import assertk.assert
import assertk.assertions.isEqualTo
import liqp.LiquidDefaults
import liqp.LiquidDefaults.defaultFilters
import liqp.LiquidParser
import liqp.Mocks
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(Parameterized::class)
data class MinusTest(val test: String, val expected: String) {
  val lcontext = Mocks.mockRenderContext()

  @Test
  fun minusTest() {
    val template = LiquidParser.newInstance().parse(test)
    val rendered = template.render()

    assert(rendered).isEqualTo(expected)
  }

  @Test
  fun invalid1() {
    assert(defaultFilters["minus"].onFilterAction(lcontext, 1))
        .isEqualTo(1)
  }

  @Test
  fun invalid2() {
    assert(
        defaultFilters["minus"].onFilterAction(lcontext, 1, 2, 3))
        .isEqualTo(-4L)
  }

  /*
   * def test_minus
   *   assert_template_result "4", "{{ input | minus:operand }}", 'input' => 5, 'operand' => 1
   *   assert_template_result "2.3", "{{ '4.3' | minus:'2' }}"
   * end
   */
  @Test
  fun originalTest() {

    assert(LiquidParser.newInstance().parse("{{ input | minus:operand }}")
        .render("{\"input\":5, \"operand\":1}")).isEqualTo("4")
    assert(LiquidParser.newInstance().parse("{{ '4.3' | minus:'2' }}")
        .render()).isEqualTo("2.3")
  }

  companion object {
    @Parameterized.Parameters(name = "{0}={1}")
    @JvmStatic
    fun tests(): Array<Array<String>> {
      return arrayOf(
          arrayOf("{{ 8 | minus: 2 }}", "6"),
          arrayOf("{{ 8 | minus: 3 }}", "5"),
          arrayOf("{{ 8 | minus: 3. }}", "5.0"),
          arrayOf("{{ 8 | minus: 3.0 }}", "5.0"),
          arrayOf("{{ 8 | minus: 3.5 }}", "4.5"),
          arrayOf("{{ 8.6 | minus: 3.5 }}", "5.1"),
          arrayOf("{{ 8 | minus: 2.0 }}", "6.0"))
    }
  }
}
